Mage.Tests/src/test/java/org/mage/test/cards/rules/AttachmentTest.java

```diff
@@ -0,0 +1,359 @@
+        addCard(Zone.GRAVEYARD, playerA, "Counterspell"); // Arbitrary instant
+
+        activateManaAbility(1, PhaseStep.PRECOMBAT_MAIN, playerA, "{T}: Add {U}.", 5);
+        castSpell(1, PhaseStep.PRECOMBAT_MAIN, playerA, "Spellweaver Volute", "Counterspell");
+        waitStackResolved(1, PhaseStep.PRECOMBAT_MAIN);
+        activateAbility(1, PhaseStep.PRECOMBAT_MAIN, playerA, "{T}: Attach", "Spellweaver Volute");
+        addTarget(playerA, codsworth);
+
+        setStopAt(1, PhaseStep.BEGIN_COMBAT);
+        execute();
+
+        assertGraveyardCount(playerA, "Spellweaver Volute", 0); // TODO: Currently goes to graveyard. Must fix
+        assertAttachedTo(playerA, "Spellweaver Volute", codsworth, false);
+    }
+
+    /**
+     * Tests that when a player gains protection from an aura, it falls off (is moved to the graveyard).
+     */
+    @Test
+    public void testCurseFallsOffFromGainedProtection() {
+        setStrictChooseMode(true);
+
+        addCard(Zone.HAND, playerA, "Curse of Opulence"); // Arbitrary curse (enchant player)
+        addCard(Zone.HAND, playerB, "Runed Halo"); // You have protection from the chosen card name.
+        addCard(Zone.BATTLEFIELD, playerA, "Mountain");
+        addCard(Zone.BATTLEFIELD, playerB, "Plains", 2);
+
+        castSpell(1, PhaseStep.PRECOMBAT_MAIN, playerA, "Curse of Opulence", playerB);
+        castSpell(2, PhaseStep.PRECOMBAT_MAIN, playerB, "Runed Halo");
+        setChoice(playerB, "Curse of Opulence");
+
+        setStopAt(2, PhaseStep.BEGIN_COMBAT);
+        execute();
+
+        assertGraveyardCount(playerA, "Curse of Opulence", 1);
+    }
+
+    /**
+     * Tests that when an aura is cast and the target becomes illegal (in this
+     * case, due to a change in card type), the aura goes to the graveyard.
+     */
+    @Test
+    public void testCastAuraIllegalTarget() {
+        setStrictChooseMode(true);
+
+        addCard(Zone.HAND, playerA, "Arcane Flight"); // Arbitrary "enchant creature" aura
+        addCard(Zone.HAND, playerA, "One with the Stars"); // Enchanted permanent loses creature type
+        addCard(Zone.BATTLEFIELD, playerA, "Runeclaw Bear"); // Arbitrary creature
+        addCard(Zone.BATTLEFIELD, playerA, "Vedalken Orrery"); // You may cast spells as though they had flash
+        addCard(Zone.BATTLEFIELD, playerA, "Island", 5);
+
+        castSpell(1, PhaseStep.PRECOMBAT_MAIN, playerA, "Arcane Flight", "Runeclaw Bear");
+        castSpell(1, PhaseStep.PRECOMBAT_MAIN, playerA, "One with the Stars", "Runeclaw Bear");
+
+        setStopAt(1, PhaseStep.BEGIN_COMBAT);
+        execute();
+
+        assertGraveyardCount(playerA, "Arcane Flight", 1);
+    }
+
+    /**
+     * Tests that when an aura tries to move from a player's hand without being cast, and
+     * it has no legal objects to attach to, it instead remains in the player's hand.
+     */
+    @Test
+    public void testAuraMoveFromHandWithNoAttachableObject() {
```

Comment:
This and below tests are missing strict choose mode

Reply:
I made the tests use strict mode but, strangely, doing so caused the (previously failing) Dream Leash test to start passing. Perhaps because it now explicitly chooses Omniscience as a target? Testing in-game revealed that it still doesn't work correctly (as in, after being sneaked into play via Show and Tell, the user is not allowed to choose an untapped permanent to attach it to, and it goes to the graveyard during state based actions). Maybe because this is a test and we can explicitly state targets, it creates an inconsistency between games / tests? Not sure, sounds like that shouldn't be the case though